File: archived/python-backend/src/mechanism/pheromone.py
from __future__ import annotations


class PheromoneTracker:
    INITIAL = 5.0
    MIN = 0.0
    MAX = 10.0
    DECAY_RATE = 0.2
    DEPOSIT_RATE = 0.5

    def __init__(self):
        self.pheromones: dict[str, float] = {}

    def get_level(self, project_id: str) -> float:
        return float(self.pheromones.get(project_id, self.INITIAL))

    def decay_all(self):
        for pid in list(self.pheromones):
            self.pheromones[pid] = self._clamp(self.pheromones[pid] * (1 - self.DECAY_RATE))

    def deposit(self, project_id: str, accuracy: float):
        current = self.get_level(project_id)
        delta = self.DEPOSIT_RATE * max(0.0, min(1.0, float(accuracy)))
        self.pheromones[project_id] = self._clamp(current + delta)

    def get_modifier(self, project_id: str) -> float:
        level = self.get_level(project_id)
        return 0.5 + (level / 10.0)

    def get_state(self) -> dict[str, float]:
        return dict(self.pheromones)

    def load_state(self, state: dict[str, float]):
        self.pheromones = {k: self._clamp(float(v)) for k, v in state.items()}

    def _clamp(self, value: float) -> float:
        return max(self.MIN, min(self.MAX, float(value)))

File: packages/backend/src/lib/pheromone.ts
/**
 * Pheromone Tracker
 * Ported from /shared/simogrants/src/mechanism/pheromone.py
 * 
 * Tracks historical accuracy of projects as "pheromone trails".
 * Projects with consistent accuracy get a funding modifier boost.
 */

const INITIAL = 5.0;
const MIN = 0.0;
const MAX = 10.0;
const DECAY_RATE = 0.2;
const DEPOSIT_RATE = 0.5;

export class PheromoneTracker {
  private pheromones: Map<string, number> = new Map();

  getLevel(projectId: string): number {
    return this.pheromones.get(projectId) ?? INITIAL;
  }

  decayAll(): void {
    for (const [pid] of this.pheromones) {
      const current = this.pheromones.get(pid)!;
      this.pheromones.set(pid, this.clamp(current * (1 - DECAY_RATE)));
    }
  }

  deposit(projectId: string, accuracy: number): void {
    const current = this.getLevel(projectId);
    const delta = DEPOSIT_RATE * Math.max(0.0, Math.min(1.0, accuracy));
    this.pheromones.set(projectId, this.clamp(current + delta));
  }

  getModifier(projectId: string): number {
    const level = this.getLevel(projectId);
    return 0.5 + level / 10.0;
  }

  getState(): Record<string, number> {
    return Object.fromEntries(this.pheromones);
  }

  loadState(state: Record<string, number>): void {
    this.pheromones = new Map(
      Object.entries(state).map(([k, v]) => [k, this.clamp(Number(v))])
    );
  }

  private clamp(value: number): number {
    return Math.max(MIN, Math.min(MAX, value));
  }
}
